File: cases/taint/kotlin/recursive-carry-negative/RecursiveCarryNegative.kt
package dataflowbench

object RecursiveCarryNegative {
    fun dfb_source(): String { // DFB-SOURCE: recursive-carry-input
        return "tainted"
    }

    fun dfb_sink(value: String) {} // DFB-SINK: recursive-carry-sink

    fun carry(value: String, depth: Int): String { // DFB-WITNESS: recursive-carry-step
        if (depth == 0) {
            return "clean"
        }
        return carry(value, depth - 1)
    }

    fun run() {
        dfb_sink(carry(dfb_source(), 5))
    }
}

File: cases/taint/c/recursive-carry-negative/recursive_carry_negative.c
int dfb_source(void) { // DFB-SOURCE: recursive-carry-input
    return 1;
}

int carry(int value, int depth) { // DFB-WITNESS: recursive-carry-step
    if (depth == 0) {
        return 0;
    }
    return carry(value, depth - 1);
}

void dfb_sink(int value) {} // DFB-SINK: recursive-carry-sink

void run(void) {
    dfb_sink(carry(dfb_source(), 5));
}
